plugin/pulp/plugin/download.py

```diff
@@ -0,0 +1,122 @@
+DEFAULT_CHUNKSIZE = 1024 * 1024
+
+
+def default_callback_generator(filename=None, size=None, checksums=None):
+    """
+    Creates a closure to write a response to disk, checksum it, and verify
+    its size.
+
+    :param filename: The absolute path the response should be saved to.
+    :param size: The size, in bytes, that the file is expected to be; check is
+                 skipped if this value is not provided.
+    :param checksums: A dictionary of checksums to validate the file with; the
+                      key should be the hash algorithm and the value the
+                      expected hex digest.
+    """
+    # TODO potentially add a way to have this generate the entire checksum set.
+    # Maybe if there's {'algorithm': None}?
+    checksums = checksums or {}
+
+    def callback(self, response):
+        """
+        A callback for the requests_futures library.
+
+        :param resp: The requests response object
+        """
+        hashes = {algorithm: hashlib.new(algorithm) for algorithm in checksums}
+        with open(filename, mode='wb') as fp:
+            chunks = response.iter_content(chunk_size=DEFAULT_CHUNKSIZE)
+            for chunk in chunks:
+                for hash in hashes.values():
+                    hash.update(chunk)
+                fp.write(chunk)
+
+        # Validate that what we downloaded is what the user expects
+        # TODO add more reasonable exceptions
+        if size and size != os.stat(filename).st_size:
+            raise ValueError()
+
+        if checksums != {algorithm: hash.hexdigest() for algorithm, hash in hashes.items()}:
+            raise ValueError()
+
+    return callback
+
+
+class FileFuturesSession(FuturesSession):
+    """
+    This class implements the requests-futures API, with one  small adjuestment. It provides
+    a default callback that writes a file to disk and validates its content. Once it
+    successfully returns, the file is ready to be imported into Pulp.
+
+        >>> future_resp = asynchronous_session.get('https://example.com')
+        >>> filename, resp = future_resp.result()
+        >>> with open(filename) as fd:
+        >>>     django_file = django.core.files.File(fd)
+        >>> artifact = pulp.apps.models.Artifact(file=django_file)
+
+    We should probably provide nice callback options to spit out Artifact objects or even Content
+    objects.
+    """
+
+    def __init__(self, executor=None, max_workers=5, session=None, *args,
+                 **kwargs):
+        super(FileFuturesSession, self).__init__(*args, **kwargs)
+
+    def request(self, *args, **kwargs):
+        """
+        This provides the requests API as well as the requests-futures API.
+
+        The only difference is a few additional keyword arguments are accepted.
+        """
+        callback_generator = kwargs.pop('callback_generator', default_callback_generator)
+        callback_generator_kwargs = kwargs.pop(
```

Comment:
It's not worth the trouble.  As a plugin writer, I'd rather just use `requests` and `futures` directly.